File: exercise/44/solution.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        vector<bool> match(p.length() + 1, false);
        match[0] = true;
        int pre, next_pre;
        for (int i = 0; i <= s.length(); i++) {
            pre = match[0];
            if (i > 0) match[0] = false;
            for (int j = 1; j <= p.length(); j++) {
                next_pre = match[j];
                if (p[j-1] != '*') {
                    match[j] = false;
                }
                if (match[j-1] && p[j-1] == '*') {
                    match[j] = true;
                }
                if (i > 0 && pre && (s[i-1] == p[j-1] || p[j-1] == '?')) {
                    match[j] = true;
                }
                pre = next_pre;
            }
        }
        return match[p.length()];
    }
};

File: exercise/44/solution.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        match = [False for i in range(len(s) + 1)]
        match[0] = True
        pre = False
        for i in range(len(s) + 1):
            pre = match[0]
            if i != 0:
                match[0] = False
            for j in range(1, len(p) + 1):
                new_pre = match[j]
                match[j] = False
                if pre and i > 0 and (s[i-1] == p[j-1] or p[j-1] == '.'):
                    match[j] = True
                elif match[j-1] and p[j-1] == '*':
                    match[j] = True
                elif i > 0 and new_pre and p[j-1] == '*':
                    match[j] = True
                pre = new_pre
        return match[p+1]


